presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java

```diff
@@ -167,4 +170,46 @@ public static FieldId checkAndGetColumnReferenceField(Expression expression, Mul
 
         return columnReferences.get(NodeRef.of(expression)).iterator().next();
     }
+
+    public static boolean isNonNullConstant(Expression expression)
+    {
+        Expression tempExpression = expression;
+        while (tempExpression instanceof Cast) {
+            tempExpression = ((Cast) tempExpression).getExpression();
+        }
+
+        if (tempExpression instanceof NullLiteral) {
+            return false;
+        }
+
+        // now allow for things like ARRAY, ROW(...) where null is OK
+        return isConstant(tempExpression);
+    }
+
+    public static boolean isConstant(Expression expression)
+    {
+        Expression tempExpression = expression;
+        while (tempExpression instanceof Cast) {
+            tempExpression = ((Cast) tempExpression).getExpression();
+        }
+
+        if (tempExpression instanceof Literal || tempExpression instanceof ArrayConstructor) {
+            return true;
+        }
+
+        // ROW an MAP are special so we explicitly do that here.
+        if (tempExpression instanceof Row) {
+            return !(((Row) tempExpression).getItems().stream().filter(x -> !isConstant(expression)).findAny().isPresent());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It is more convenient and readable to directly use Stream::anyMatch instead of Stream::filter, Stream::findAny, and Optional::isPresent when the matched data returned by Stream::findAny is not used anywhere else.